src/common/src/array/data_chunk.rs

```diff
@@ -322,83 +320,24 @@ impl DataChunk {
+        };
 
-        let mut total_capacity = chunks.iter().map(|chunk| chunk.capacity()).sum();
-        let num_chunks = (total_capacity + each_size_limit - 1) / each_size_limit;
+        let mut builder = DataChunkBuilder::new(data_types, each_size_limit);
+        let mut outputs = Vec::new();
 
-        // the idx of `chunks`
-        let mut chunk_idx = 0;
-        // the row idx of `chunks[chunk_idx]`
-        let mut start_row_idx = 0;
-        // how many rows does this new chunk need?
-        let mut new_chunk_require = std::cmp::min(total_capacity, each_size_limit);
-        let mut array_builders: Vec<ArrayBuilderImpl> = chunks[0]
-            .columns
-            .iter()
-            .map(|col| col.create_builder(new_chunk_require))
-            .collect();
-        let mut array_len = new_chunk_require;
-        let mut new_chunks = Vec::with_capacity(num_chunks);
-        while chunk_idx < chunks.len() {
-            let capacity = chunks[chunk_idx].capacity();
-            let num_rows_left = capacity - start_row_idx;
-            let actual_acquire = std::cmp::min(new_chunk_require, num_rows_left);
-            let end_row_idx = start_row_idx + actual_acquire - 1;
-            array_builders
-                .iter_mut()
-                .zip_eq_fast(chunks[chunk_idx].columns())
-                .for_each(|(builder, column)| {
-                    let mut array_builder = column.create_builder(end_row_idx - start_row_idx + 1);
-                    for row_idx in start_row_idx..=end_row_idx {
-                        array_builder.append(column.value_at(row_idx));
-                    }
-                    builder.append_array(&array_builder.finish());
-                });
-            // since `end_row_idx` is inclusive, exclude it for the next round.
-            start_row_idx = end_row_idx + 1;
-            // if the current `chunks[chunk_idx] is used up, move to the next one
-            if start_row_idx == capacity {
-                chunk_idx += 1;
-                start_row_idx = 0;
-            }
-            new_chunk_require -= actual_acquire;
-            total_capacity -= actual_acquire;
-            // a new chunk receives enough rows, finalize it
-            if new_chunk_require == 0 {
-                let new_columns: Vec<ArrayRef> = array_builders
-                    .drain(..)
-                    .map(|builder| builder.finish().into())
-                    .collect();
-
-                array_builders = new_columns
-                    .iter()
-                    .map(|col_type| col_type.create_builder(new_chunk_require))
-                    .collect();
-
-                let data_chunk = DataChunk::new(new_columns, array_len);
-                new_chunks.push(data_chunk);
-
-                new_chunk_require = std::cmp::min(total_capacity, each_size_limit);
-                array_len = new_chunk_require;
+        for chunk in chunks {
+            for output in builder.append_chunk(chunk.clone()) {
```

Comment:
```suggestion
            if let Some(output) = builder.append_chunk(chunk.clone()) {
```

nit: I prefer `if let` here, `for` seems to signal that this can happen multiple times, it's more general than `if let`.

Reply:
Appending a large chunk may result in multiple smaller outputs. 😕 So the return type of `append_chunk` should be an iterator.